### PR TITLE
Feat/on chain json validation/111

### PR DESCRIPTION
**Pull request by mehtaishita**
**Problem**

We need to provide lightweight on-chain validation for Json schemas

Closes: #111 

**Solution**

Use serde-json-core and provide json validation support using no-std library

**Goals**

- [ ] Proper imports of serde-json-core
- [ ] Helper validate json schemas using serde-json-core 
- [ ] Validate schemas when registering schemas in Schema pallet
- [x] Write new tests and amend current tests to provide full coverage

**Changes made**

- Update Cargo.toml with no-std version of serde-json (serde-json-core)
- Currently using serde-json since serde-json-core is failing to compile (will update later)
- Helper class to validate json schemas
- Add json-validation in schema pallet in register schema extrinsic
- InvalidSchema Error thrown when validation fails

@enddynayn @wilwade

### CODE REVIEW

**Comment by mehtaishita** [comment]
**Working on tests, will finish today**

**Will replace serde-json-core when the external library compiles**

File: designdocs/provider_permissions.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
+# Delegator <-> Provider Grants
+
+## Context and Scope
```
Comment (shannonwells) [inline designdocs/provider_permissions.md:3]: I don’t think this should be in this PR

File: pallets/schemas/src/tests.rs [hunk 2]
```diff
@@ -3,16 +3,24 @@ use common_primitives::schema::SchemaId;
...
 use sp_runtime::DispatchError::BadOrigin;
+use common_helpers::serde;
 
 use super::mock::*;
 
 fn create_bounded_schema_vec(
 	from_string: &str,
 ) -> BoundedVec<u8, <Test as Config>::SchemaMaxBytesBoundedVecLimit> {
 	let fields_vec = Vec::from(from_string.as_bytes());
+	println!("Bounded vector convereted in helper function: {:?}",fields_vec);
```
Comment (shannonwells) [inline pallets/schemas/src/tests.rs:14]: This looks like a left over debug print

File: pallets/schemas/src/tests.rs [hunk 3]
```diff
@@ -160,3 +170,23 @@ fn get_non_existing_schema_by_id_should_return_none() {
...
+#[test]
+fn reject_null_json_schema() {
+	new_test_ext().execute_with(|| {
+		assert_noop!(SchemasPallet::ensure_valid_schema(
+			create_bounded_schema_vec("null")
+		), Error::<Test>::InvalidSchema);
+	})
+}
+
+
```
Comment (shannonwells) [inline pallets/schemas/src/tests.rs:192]: This needs a malformed JSON negative test, even a simple one although it’s kind of testing the serde parser

**Review by shannonwells — COMMENTED** [review]
Looks ok, although I can’t test it here. There are commented out lines that should be removed.

File: common/helpers/src/serde.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// use serde_json_core::{Result, Value, from_slice};
+use serde_json::{Result, Value, from_slice};
+
+pub fn validate_json_schema(json_schema: Vec<u8>) -> Result<()> { // short curcuit
+    let result: Value = from_slice(&json_schema)?;
```
Comment (aramikm) [inline common/helpers/src/serde.rs:5]: Does this support json with UTF-8 input? Maybe we need a test with UTF-8 data to check.

File: common/helpers/Cargo.toml [hunk 5]
```diff
@@ -17,6 +17,8 @@ apache-avro = { git = "https://github.com/apache/avro", features = ["snappy"] }
 thiserror = "1"
 sha2 = "0.10.2"
 rayon = "1.5"
+# serde_json_core = { version = "0.3" , default-features = false }
+serde_json = { version = "1.0", default-features = false }
```
Comment (aramikm) [inline common/helpers/Cargo.toml:21]: shouldn't this be the following based on https://crates.io/crates/serde_json#no-std-support?

`serde_json = { version = "1.0", default-features = false, features = ["alloc"] }`

File: pallets/schemas/src/tests.rs [hunk 6]
```diff
@@ -134,7 +142,9 @@ fn get_existing_schema_by_id_should_return_schema() {
 		let sender: AccountId = 1;
 		sudo_set_max_schema_size();
 		// arrange
-		let test_str = "foo,bar,bazz";
+		// let test_str = "foo,bar,bazz";
```
Comment (aramikm) [inline pallets/schemas/src/tests.rs:145]: seems leftover

**Review by aramikm — COMMENTED** [review]
I've added some questions and comments.

File: designdocs/provider_permissions.md (hunk 1 above)
Comment (mehtaishita) [reply designdocs/provider_permissions.md:3]: it shouldn't be, but these changes show up as merge conflicts when I try to merge my branch (I haven't made any changes to this file other than merging main into my branch

File: common/helpers/Cargo.toml (hunk 5 above)
Comment (mehtaishita) [reply common/helpers/Cargo.toml:21]: Yes, so I'll create another pr for using serde-json-core. currently that library is failing to compile

File: common/helpers/src/serde.rs (hunk 4 above)
Comment (mehtaishita) [reply common/helpers/src/serde.rs:5]: So I would test with Json with utf-8 encoding, and the get-schema-id should return the proper schema right?

**Comment by mehtaishita** [comment]
See https://github.com/LibertyDSNP/frequency/pull/216 for updated pull request

**Pull request closed without merging**